### PR TITLE
Warn if git version is not high enough for supporting all features

### PR DESCRIPTION
**Pull request by sigurdm**
Fixes: https://github.com/dart-lang/pub/issues/3275

### CODE REVIEW

File: lib/src/git.dart [hunk 1]
```diff
@@ -121,13 +122,32 @@ String? repoRoot(String dir) {
...
+    if (match == null) return false;
+    // Git seems to use many parts in the version number. We just check the
+    // first two.
+    final major = int.parse(match[1]!);
+    final minor = int.parse(match[2]!);
+    if (Version(major, minor, 0) < _minSupportedGitVersion) {
+      // We just warn here, as some features might work with older versions of
+      // git.
+      log.warning('''
+Pub need at least git version 2.14 for all features to work.
```
Comment (szakarias) [inline lib/src/git.dart:146]: need -> needs

**Review by szakarias — APPROVED** [review]

File: lib/src/git.dart (hunk 1 above)
Comment (sigurdm) [reply lib/src/git.dart]: Thanks!

File: lib/src/git.dart [hunk 2]
```diff
@@ -121,13 +122,32 @@ String? repoRoot(String dir) {
   return null;
 }
 
+final _minSupportedGitVersion = Version(2, 14, 0);
```
Comment (jonasfj) [inline lib/src/git.dart:125]: This needs a documentation string that explains what features were adding in Git 2.14 that causes it to the be the minimum supported version.

Comment (jonasfj) [reply lib/src/git.dart:127]: We only used that in:
```sh
$ git grep recurse-submodules
lib/src/validator/gitignore.dart:        '--recurse-submodules'
```
So I'm guessing this is actually a problem we've caused, one could even call it a regression.

File: lib/src/git.dart [hunk 3]
```diff
@@ -121,13 +122,34 @@ String? repoRoot(String dir) {
...
+    // Git seems to use many parts in the version number. We just check the
+    // first two.
+    final major = int.parse(match[1]!);
+    final minor = int.parse(match[2]!);
+    if (Version(major, minor, 0) < _minSupportedGitVersion) {
+      // We just warn here, as some features might work with older versions of
+      // git.
+      log.warning('''
+Pub needs at least git version 2.14 for all features to work.
+You seem to have ${output.substring('git '.length)}.
```
Comment (jonasfj) [inline lib/src/git.dart:149]: We have a variable for topLevelCommand... 
```suggestion
You have a very old version of git (version ${output.substring('git '.length)}),
for $topLevelCommand it is recommanded to use git version 2.14 or newer.
```

File: lib/src/git.dart (hunk 2 above)
Comment (sigurdm) [reply lib/src/git.dart:128]: This was used in Package.listFiles before that. So the regression is not introduced at that point.

**Review by jonasfj — APPROVED** [review]

**Pull request merged**